File: windows/SettingsStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text.Json;

namespace Appreciate
{
    /// <summary>
    /// Persists user preferences to %APPDATA%/Appreciate/settings.json.
    /// Default packs are loaded from common/packs.json bundled alongside the exe.
    /// </summary>
    public class SettingsStore
    {
        private static readonly string SettingsDir =
            Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "Appreciate");
        private static readonly string SettingsFile = Path.Combine(SettingsDir, "settings.json");

        private static SettingsStore? _instance;
        public static SettingsStore Instance => _instance ??= Load();

        /// <summary>Loaded from packs.json at startup.</summary>
        private static readonly (Dictionary<string, string> Packs, string DefaultPack) Bundled = LoadBundledPacks();

        // In-memory cache for remote packs
        private static readonly Dictionary<string, List<string>> _remoteCache = new();
        private static readonly Dictionary<string, DateTime> _remoteFetchTime = new();
        private static readonly TimeSpan _cacheDuration = TimeSpan.FromHours(1);
        private static readonly HttpClient _httpClient = new() { Timeout = TimeSpan.FromSeconds(10) };

        private static (Dictionary<string, string>, string) LoadBundledPacks()
        {
            try
            {
                // Look for packs.json next to the executable
                var exeDir = AppDomain.CurrentDomain.BaseDirectory;
                var path = Path.Combine(exeDir, "packs.json");
                if (!File.Exists(path))
                    path = Path.Combine(exeDir, "..", "common", "packs.json");
                if (!File.Exists(path))
                    path = Path.Combine(exeDir, "..", "..", "..", "common", "packs.json");

                if (File.Exists(path))
                {
                    var json = File.ReadAllText(path);
                    using var doc = JsonDocument.Parse(json);
                    var root = doc.RootElement;
                    var defaultPack = root.TryGetProperty("default_pack", out var dp) ? dp.GetString() ?? "" : "";
                    var packs = new Dictionary<string, string>();
                    if (root.TryGetProperty("packs", out var packsEl))
                    {
                        foreach (var prop in packsEl.EnumerateObject())
                        {
                            var lines = new List<string>();
                            foreach (var line in prop.Value.EnumerateArray())
                                lines.Add(line.GetString() ?? "");
                            packs[prop.Name] = string.Join("\n", lines);
                        }
                    }
                    return (packs, defaultPack);
                }
            }
            catch { }
            return (new Dictionary<string, string>(), "");
        }

        /// <summary>User's packs (editable dictionary).</summary>
        public Dictionary<string, string> Packs { get; set; } = new(Bundled.Packs);
        public string SelectedPack { get; set; } = Bundled.DefaultPack;
        public float MinIntervalMinutes { get; set; } = 0.1f;
        public float MaxIntervalMinutes { get; set; } = 1.5f;
        public float DisplayDurationSeconds { get; set; } = 6f;
        public bool IsEnabled { get; set; } = true;
        public bool LaunchAtLogin { get; set; } = true;

        /// <summary>The reminder text of the current pack.</summary>
        public string ReminderText
        {
            get => Packs.TryGetValue(SelectedPack, out var text) ? text : "";
            set { Packs[SelectedPack] = value; }
        }

        /// <summary>Sorted pack names for display.</summary>
        public string[] PackNames => Packs.Keys.OrderBy(k => k).ToArray();

        /// <summary>Add a new pack. Returns false if name already exists.</summary>
        public bool AddPack(string name)
        {
            if (string.IsNullOrWhiteSpace(name) || Packs.ContainsKey(name)) return false;
            Packs[name] = "";
            SelectedPack = name;
            Save();
            return true;
        }

        /// <summary>Delete a pack. Cannot delete last remaining pack.</summary>
        public bool DeletePack(string name)
        {
            if (Packs.Count <= 1 || !Packs.ContainsKey(name)) return false;
            Packs.Remove(name);
            if (SelectedPack == name)
                SelectedPack = Packs.Keys.OrderBy(k => k).First();
            Save();
            return true;
        }

        /// <summary>Returns a random line from ReminderText.</summary>
        public string RandomLine
        {
            get
            {
                var text = ReminderText;
                if (IsRemotePack(text))
                {
                    var lines = FetchRemotePack(text.Trim());
                    return lines.Count > 0 ? lines[Random.Shared.Next(lines.Count)] : "";
                }
                var localLines = text.Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
                return localLines.Length > 0 ? localLines[Random.Shared.Next(localLines.Length)] : text;
            }
        }

        private static bool IsRemotePack(string text)
        {
            var trimmed = text.Trim();
            return trimmed.StartsWith("https://") && !trimmed.Contains('\n');
        }

        private static List<string> FetchRemotePack(string url)
        {
            if (_remoteCache.TryGetValue(url, out var cached))
            {
                if (_remoteFetchTime.TryGetValue(url, out var fetchTime))
                {
                    if (DateTime.Now - fetchTime < _cacheDuration)
                        return cached;
                }
            }

            try
            {
                var text = _httpClient.GetStringAsync(url).Result;
                var lines = text.Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries).ToList();
                _remoteCache[url] = lines;
                _remoteFetchTime[url] = DateTime.Now;
                return lines;
            }
            catch { }

            return _remoteCache.TryGetValue(url, out var fallback) ? fallback : new List<string>();
        }

        public void Save()
        {
            Directory.CreateDirectory(SettingsDir);
            var json = JsonSerializer.Serialize(this, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(SettingsFile, json);
        }

        private static SettingsStore Load()
        {
            try
            {
                if (File.Exists(SettingsFile))
                {
                    var json = File.ReadAllText(SettingsFile);
                    var store = JsonSerializer.Deserialize<SettingsStore>(json) ?? new SettingsStore();
                    // Ensure selected pack exists
                    if (!store.Packs.ContainsKey(store.SelectedPack))
                        store.SelectedPack = store.Packs.Keys.OrderBy(k => k).FirstOrDefault() ?? "";
                    return store;
                }
            }
            catch { }
            return new SettingsStore();
        }
    }
}

File: macos/Sources/SettingsStore.swift
import Foundation
import Combine
import ServiceManagement

/// Persists user preferences via UserDefaults.
final class SettingsStore: ObservableObject {
    static let shared = SettingsStore()

    private let defaults = UserDefaults.standard

    // In-memory cache for remote packs
    private static var remoteCache: [String: [String]] = [:]
    private static var remoteFetchTime: [String: Date] = [:]
    private static let cacheDuration: TimeInterval = 3600 // 1 hour

    private enum Keys {
        static let packs = "packs"
        static let selectedPack = "selectedPack"
        static let minIntervalMinutes = "minIntervalMinutes"
        static let maxIntervalMinutes = "maxIntervalMinutes"
        static let displayDurationSeconds = "displayDurationSeconds"
        static let isEnabled = "isEnabled"
        static let launchAtLogin = "launchAtLogin"
    }

    /// Reads default packs from the bundled packs.json (common/packs.json).
    private static func loadBundledPacks() -> (packs: [String: String], defaultPack: String) {
        guard let url = Bundle.main.url(forResource: "packs", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let packsDict = json["packs"] as? [String: [String]] else {
            return (packs: [:], defaultPack: "")
        }
        let defaultPack = json["default_pack"] as? String ?? ""
        var result: [String: String] = [:]
        for (name, lines) in packsDict {
            result[name] = lines.joined(separator: "\n")
        }
        return (packs: result, defaultPack: defaultPack)
    }

    /// The built-in packs loaded from packs.json.
    static let bundled = loadBundledPacks()

    /// User's packs (editable dictionary). Keys = pack names, values = newline-separated reminder text.
    @Published var packs: [String: String] {
        didSet { savePacks() }
    }

    @Published var selectedPack: String {
        didSet {
            defaults.set(selectedPack, forKey: Keys.selectedPack)
            objectWillChange.send()
        }
    }

    @Published var minIntervalMinutes: Double {
        didSet { defaults.set(minIntervalMinutes, forKey: Keys.minIntervalMinutes) }
    }

    @Published var maxIntervalMinutes: Double {
        didSet { defaults.set(maxIntervalMinutes, forKey: Keys.maxIntervalMinutes) }
    }

    @Published var displayDurationSeconds: Double {
        didSet { defaults.set(displayDurationSeconds, forKey: Keys.displayDurationSeconds) }
    }

    @Published var isEnabled: Bool {
        didSet { defaults.set(isEnabled, forKey: Keys.isEnabled) }
    }

    @Published var launchAtLogin: Bool {
        didSet {
            defaults.set(launchAtLogin, forKey: Keys.launchAtLogin)
            updateLoginItem()
        }
    }

    /// The reminder text of the currently selected pack.
    var reminderText: String {
        get { packs[selectedPack] ?? "" }
        set { packs[selectedPack] = newValue }
    }

    /// Sorted pack names for display.
    var packNames: [String] {
        packs.keys.sorted()
    }

    private init() {
        let bundled = SettingsStore.bundled

        defaults.register(defaults: [
            Keys.selectedPack: bundled.defaultPack,
            Keys.minIntervalMinutes: 0.1,
            Keys.maxIntervalMinutes: 1.5,
            Keys.displayDurationSeconds: 6.0,
            Keys.isEnabled: true,
            Keys.launchAtLogin: true,
        ])

        // Load packs from UserDefaults or use bundled defaults
        if let data = defaults.data(forKey: Keys.packs),
           let saved = try? JSONDecoder().decode([String: String].self, from: data) {
            self.packs = saved
        } else {
            self.packs = bundled.packs
        }

        self.selectedPack = defaults.string(forKey: Keys.selectedPack) ?? bundled.defaultPack
        self.minIntervalMinutes = defaults.double(forKey: Keys.minIntervalMinutes)
        self.maxIntervalMinutes = defaults.double(forKey: Keys.maxIntervalMinutes)
        self.displayDurationSeconds = defaults.double(forKey: Keys.displayDurationSeconds)
        self.isEnabled = defaults.bool(forKey: Keys.isEnabled)
        self.launchAtLogin = defaults.bool(forKey: Keys.launchAtLogin)

        // Ensure selected pack exists
        if packs[selectedPack] == nil {
            selectedPack = packs.keys.sorted().first ?? ""
        }

        updateLoginItem()
    }

    private func savePacks() {
        if let data = try? JSONEncoder().encode(packs) {
            defaults.set(data, forKey: Keys.packs)
        }
    }

    /// Add a new pack with the given name. Returns false if name already exists.
    func addPack(name: String) -> Bool {
        guard !name.isEmpty, packs[name] == nil else { return false }
        packs[name] = ""
        selectedPack = name
        return true
    }

    /// Delete a pack. Cannot delete the last remaining pack.
    func deletePack(name: String) -> Bool {
        guard packs.count > 1, packs[name] != nil else { return false }
        packs.removeValue(forKey: name)
        if selectedPack == name {
            selectedPack = packs.keys.sorted().first ?? ""
        }
        return true
    }

    private func updateLoginItem() {
        let service = SMAppService.mainApp
        do {
            if launchAtLogin {
                try service.register()
            } else {
                try service.unregister()
            }
        } catch {
            NSLog("[Appreciate] Login item update failed: %@", error.localizedDescription)
        }
    }

    /// Returns a random line from the current pack's reminder text.
    var randomLine: String {
        let text = reminderText
        if isRemotePack(text) {
            let lines = fetchRemotePack(url: text.trimmingCharacters(in: .whitespacesAndNewlines))
            return lines.randomElement() ?? ""
        }
        let lines = text.components(separatedBy: "\n").map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
        return lines.randomElement() ?? text
    }

    private func isRemotePack(_ text: String) -> Bool {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.hasPrefix("https://") && !trimmed.contains("\n")
    }

    private func fetchRemotePack(url: String) -> [String] {
        // Check cache
        if let cached = Self.remoteCache[url],
           let fetchTime = Self.remoteFetchTime[url],
           Date().timeIntervalSince(fetchTime) < Self.cacheDuration {
            return cached
        }

        // Fetch synchronously
        guard let urlObj = URL(string: url) else { return Self.remoteCache[url] ?? [] }
        var request = URLRequest(url: urlObj)
        request.timeoutInterval = 10
        request.httpMethod = "GET"

        do {
            let data = try Data(contentsOf: urlObj, options: .alwaysMapped)
            if let text = String(data: data, encoding: .utf8) {
                let lines = text.components(separatedBy: "\n")
                    .map { $0.trimmingCharacters(in: .whitespaces) }
                    .filter { !$0.isEmpty }
                Self.remoteCache[url] = lines
                Self.remoteFetchTime[url] = Date()
                return lines
            }
        } catch {
            NSLog("[Appreciate] Failed to fetch remote pack: %@", error.localizedDescription)
        }

        return Self.remoteCache[url] ?? []
    }
}
